File: unpack.f90
module unpack

  implicit none

  real, allocatable :: rdata(:, :)
  character(1), allocatable :: cdata(:)

  character(80) :: arl
  character(50) :: header_label 
  character(3072) :: header

  integer :: year, month, day, hour
  integer :: ifc
  character(4) :: var_desc

  character(4) :: model
  integer :: grid_num, z_coord
  real :: pole_lat, pole_lon, ref_lat
  real :: ref_lon, grid_size, orient
  real :: tan_lat, sync_x, sync_y
  real :: sync_lat, sync_lon, reserved
  integer :: nx, ny, nz

  integer :: nxy
  integer :: rec_len

  integer :: rstat, cstat, iostat

  integer :: i, k
  integer :: nexp 
  real :: var1, prec

  interface
    ! unpacked character data is resolved to real_data 
    subroutine decode(cdata, rdata, nx, ny, nexp, var1)
        character(1), intent(in) :: cdata(:)
        real, intent(out) :: rdata(:, :)
        integer, intent(in) :: nx, ny, nexp
        real, intent(in) :: var1
    end subroutine
  end interface

  write(*,*) 'Enter ARL file name:'
  read(*, '(a)') arl
  arl = adjustl(arl)

  open(10, file = arl, recl = 150, access = 'direct', form = 'unformatted')
  read(10, rec = 1) header_label, header(1:108)
  read(header_label,'(5i2,4x,a4)') year, month, day, hour, ifc, var_desc
  !! Debug
  ! write(*, '(a, 4i2)') 'YYMMDDHH: ', year, month, day, hour
  read(header(1:108), '(a4, 1i3, 1i2,  12f7.0, 3i3)')         & 
       model, grid_num, z_coord,                              &
       pole_lat, pole_lon, ref_lat,                           &
       ref_lon, grid_size, orient,                            &
       tan_lat, sync_xp, sync_yp,                             &
       sync_lat, sync_lon, reserved,                          &
       nx, ny, nz  
  close(10)

  nxy = nx * ny
  rec_len = nxy + 50
  
  allocate(real_data(nx,ny), stat = rstat)
  allocate(char_data(nxy), stat = cstat)

  open(10, file = arl, recl = rec_len, acces = 'direct', form = 'unformatted')

  i = 1
  do 
    read(10, rec = i, iostat = iostat) label, (cdata(k), k = 1, nxy)
    if ( iostat > 0 ) then 
      exit
    else 
      read(label, '(6i2, 2x, a4, 1i4, 2f7.0)') & 
      year, month, day, hour, grid_num, level, & 
      var_desc, nexp, prec, var1
      ! unpack data from header 
      if(var_desc /= 'INDX')  then 
        ! call decode(cdata, rdata, nx, ny, nexp, var1)   
      end if
    end if 
    i = i + 1
  end do

end module

! subroutine decode(cdata, rdata, nx, ny, nexp, var1) 

!     character(1), intent(in) :: cdata(:)    
!     real, intent(out) :: rdata(:, :)
!     integer, intent(in) :: nx, ny, nexp
!     real, intent(in) :: var1
!     real :: scale
!     real :: vold
!     integer :: indx
!     integer :: i, j

!     scale = 2.0 ** (7 - nexp)
!     vold = var1 
!     indx = 0

!     do j = 1, ny
!         do i = 1, nx 
!             indx = indx + 1
!             rdata(i,j) = (ichar(cdata(indx)) - 127.0) / scale + vold
!             vold = rdata(i,j)
!         end do
!         vold = rdata(1,j)
!     end do

! end subroutine

File: unpack.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>
#include <netcdf.h>
#include <cmapf.h>

#define LABSIZE 50

int numX(char str[]);
int numY(char str[]);
char *varDesc(char str[]);
long numRecs(long size, long recl);
double numExp(char str[]);
double numVar1(char str[]);
void unpack(double nexp, double var1, size_t nx, size_t ny,
            char cdata[], double rdata[nx][ny]);

int main(int argc, char *argv[])
{

    FILE *arl;  // ARL file stream
    long fsize; // file byte size

    char *label;  // standard label 50-bytes
    char *header; // header 108-bytes
    char *hindex; // header index desc

    int nx, ny;     // x, y dims
    long nxy, recl; // num xy, record length
    double nexp;    // scaling integer
    double var1;    // variable at (1,1)
    
    // Cmapf
    maparam stcprm; // Map parameter struct

    //stcm1p()


    // Open ARL to file stream
    arl = fopen(argv[1], "rb");

    // Get the file size
    fseek(arl, 0, SEEK_END);
    fsize = ftell(arl);
    rewind(arl);

    // Allocate label and ARL header array space
    label = malloc(sizeof(char) * LABSIZE);
    header = malloc(sizeof(char) * 3072); // Overflow

    // Read the standard portion of the label(50) and header(108)
    fread(label, sizeof(char), LABSIZE, arl);
    fread(header, sizeof(char), 108, arl);

    puts(header); // print header

    // Header INDX label
    hindex = varDesc(label);

    // Get grid dimensions & record length
    nx = numX(header);
    ny = numY(header);
    nxy = nx * ny;
    recl = nxy + LABSIZE;

    // Define data arrays
    char *cdata;                            // packed 1-d char array
    double(*rdata)[(size_t)nx][(size_t)ny]; // unpacked 2-d real array

    // Allocate array space
    cdata = malloc(sizeof(char) * ((size_t)nxy));
    rdata = malloc(sizeof(double[(size_t)nx][(size_t)ny]));

    // Unpack each record described by its 50-byte record label
    rewind(arl);
    int rec = 0;
    while (rec < numRecs(fsize, recl))
    {
        ++rec;
        fread(label, sizeof(char), LABSIZE, arl);
        fread(cdata, sizeof(char), (size_t)nxy, arl);
        if (strcmp(varDesc(label), hindex) != 0)
        {
            nexp = numExp(label);
            var1 = numVar1(label);
            printf("\n%s\n", label);
            unpack(nexp, var1, (size_t)nx, (size_t)ny, cdata, *rdata);
        }
    }

    // Close ARL file stream
    fclose(arl);

    return 0;
}

void unpack(double nexp, double var1, size_t nx, size_t ny,
            char cdata[], double rdata[nx][ny])
{
    int indx = 0;
    double vold = var1;
    double scale = pow(2.0, (7.0 - nexp));

    for (size_t j = 0; j < ny; ++j)
    {
        for (size_t i = 0; i < nx; ++i)
        {
            rdata[i][j] = (cdata[indx] - 127.0) / scale + vold;
            vold = rdata[i][j];
            ++indx;
            if (i < 5 && j < 25)
                printf("(%ld,%ld): %lf ", i, j, rdata[i][j]);
        }
        vold = rdata[0][j];
        if (j < 25)
            printf(" ...\n");
    }
}

// Be simple.
// Pull string from header
char *pullStr(char str[], int pos, int len)
{   
    char *sptr;
    sptr = malloc(len + 1);
    strncpy(sptr, str + pos, len); 
    sptr[len] = '\0';
    return sptr;
}

int numX(char str[])
{
    int ix;
    char *cx;

    cx = pullStr(str, 93, 3);
    ix = atoi(cx);

    return ix;
}

int numY(char str[])
{
    int iy;
    char *cy;

    cy = pullStr(str, 96, 3);
    iy = atoi(cy);

    return iy;
}

char *varDesc(char str[])
{
    char *desc;

    desc = pullStr(str, 14, 4);

    return desc;
}

long numRecs(long size, long recl)
{
    return size / recl;
}

double numExp(char str[])
{
    int exp;
    char *nexp; 

    nexp = pullStr(str, 18, 4);    
    exp = atof(nexp);

    return exp;
}

double numVar1(char str[])
{
    char *var1; 
    double var;

    var1 = pullStr(str, 36, 17);
    var = atof(var1);

    return var;
}
